File: ios/Classes/style/ButtonStyle.swift
//
//  ButtonStyle.swift
//  exponea
//
//  Created by Adam Mihalik on 02/03/2023.
//

import Foundation
import UIKit

class ButtonStyle {
    var textOverride: String?
    var textColor: String?
    var backgroundColor: String?
    var showIcon: String?
    var textSize: String?
    var enabled: Bool?
    var borderRadius: String?
    var textWeight: String?

    init(
        textOverride: String? = nil,
        textColor: String? = nil,
        backgroundColor: String? = nil,
        showIcon: String? = nil,
        textSize: String? = nil,
        enabled: Bool? = nil,
        borderRadius: String? = nil,
        textWeight: String? = nil
    ) {
        self.textOverride = textOverride
        self.textColor = textColor
        self.backgroundColor = backgroundColor
        self.showIcon = showIcon
        self.textSize = textSize
        self.enabled = enabled
        self.borderRadius = borderRadius
        self.textWeight = textWeight
    }

    func applyTo(_ target: UIButton) {
        if let showIcon = showIcon {
            target.setImage(
                UIImage.parse(showIcon)?.withRenderingMode(.alwaysTemplate),
                for: .normal
            )
            target.imageView?.contentMode = .scaleAspectFit
            target.imageEdgeInsets = UIEdgeInsets(top: 4, left: 0, bottom: 4, right: 0)
        }
        if let textOverride = textOverride {
            target.setTitle(textOverride, for: .normal)
        }
        if let textColor = UIColor.parse(textColor) {
            target.setTitleColor(textColor, for: .normal)
            target.tintColor = textColor
        }
        if let backgroundColor = UIColor.parse(backgroundColor) {
            target.backgroundColor = backgroundColor
        }
        if let textSize = CGFloat.parse(textSize),
           let titleLabel = target.titleLabel {
            titleLabel.font = titleLabel.font.withSize(textSize)
        }
        if let textWeight = UIFont.Weight.parse(textWeight),
           let titleLabel = target.titleLabel {
            titleLabel.font = UIFont.systemFont(ofSize: titleLabel.font.pointSize, weight: textWeight)
        }
        if let enabled = enabled {
            target.isEnabled = enabled
        }
        if let borderRadius = CGFloat.parse(borderRadius) {
            target.layer.cornerRadius = borderRadius
        }
    }
}

File: lib/src/data/model/button_style.dart
import '../util/Codable.dart';

class SimpleButtonStyle extends Encodable {
  final String? textOverride;
  final String? textColor;
  final String? backgroundColor;
  final bool? showIcon;
  final String? textSize;
  final bool? enabled;
  final String? borderRadius;
  final String? textWeight;

  SimpleButtonStyle({
      this.textOverride,
      this.textColor,
      this.backgroundColor,
      this.showIcon,
      this.textSize,
      this.enabled,
      this.borderRadius,
      this.textWeight});

  @override
  String toString() {
    return 'ButtonStyle{textOverride: $textOverride, textColor: $textColor, backgroundColor: $backgroundColor, showIcon: $showIcon, textSize: $textSize, enabled: $enabled, borderRadius: $borderRadius, textWeight: $textWeight}';
  }

  @override
  Map<String, dynamic> encode() {
    return {
      'textOverride': textOverride,
      'textColor': textColor,
      'backgroundColor': backgroundColor,
      'showIcon': showIcon,
      'textSize': textSize,
      'enabled': enabled,
      'borderRadius': borderRadius,
      'textWeight': textWeight
    };
  }
}
